### DISCUSSION TITLE
I got an error

### DISCUSSION THREAD

**Discussion by PseudoSurfer** [haltakov/simple-photo-gallery#107] [Q&A]
Hello. I'm trying to use your Simple-Photo-Gallery to add my Google Photos to my GitHub Pages site. I haven't gotten to the point of figuring out how to add it to my Jekyll set-up. I was testing it out when I ran into a problem. The init process went fine. After giving the build command I got this error
Building the Simple Photo Gallery...
Generating thumbnails...
Generating the images_data.json file...
Starting Firefox webdriver...
Something went wrong while generating the images_data.json file: Message: Unable to find a matching set of capabilities
It stopped running and it hasn't created the HTML file yet either.
Any help would be appreciated.

**Comment by haltakov** [comment]
This is admittedly one of the less tested features of the gallery... Which OS are you using? Do you have Firefox installed?

**Reply by PseudoSurfer** [reply]
I'm using Mac OS and I didn't have Firefox installed but after seeing your question I installed it and everything went perfectly.
I'm trying to add my google photos to my Jekyll site hosted on Github.
Thanks for your help.

**Reply by haltakov** [reply]
OK, I'm happy that it works for you now, but still my setup is not really optimal... I have to find a better way to integrate that.

Let me know if you want some help integrating into Jekyll. The gallery generates static files in the end so it will surely be possible!

**Reply by PseudoSurfer** [reply]
Thanks again it works perfectly. I used the static files that your script generates. Jekyll needs to have templates set up to avoid repetition so I broke the static files down and relinked all the required assets. I have it almost perfect. (pseudosurfer.com) If you take a look at the photo galleries, you'll see that everything is fine until you open an album. I can't seem to get rid of all the white spaces around the elements.

**Reply by haltakov** [reply]
Nice integration! :) 

I looked at your code (for example https://pseudosurfer.com/photos/blyde_river_canyon.html). Simple Photo Gallery uses Bootstrap as a CSS framework. The problem is that in the code, you include it (at line 21) after you do the styling of the page (at line 10).

Move the line `<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/...` before line 10 (`<link rel="stylesheet" href="Blyde_river_canyon_files/styles.css">`) and it should be fine.

**Reply by PseudoSurfer** [reply]
After I accidentally spilt liquid on my MacBook Pro and it died, I've bought a new M1 Mac Mini. Installing Python and the required pieces for your script is proving difficult. The error message I'm getting is: 

The conflict is caused by:
    simple-photo-gallery 1.5.0 depends on opencv-python>=4.2.0.32

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

I'm not comfortable changing your script by myself. Can you suggest what I might do?

**Reply by PseudoSurfer** [reply]
I've installed the latest Python 3.9.4 using pyenv on homebrew which means I could actually install and use a previous version that works, I'm just not sure which version to use.

**Reply by PseudoSurfer** [reply]
I've managed to install all but one of the required packages albeit different versions 'opencv-python>=4.5.2.arm64', 'pillow>=8.0.1', 'selenium>=3.141.0', 'requests>=2.25.0'. The only one I can't install now is jinja2. I still have no idea how to install your simple-photo-gallery using the versions that I managed to install. I realise that this is the wrong way to about doing this but I really like how your script works.

**Reply by haltakov** [reply]
I've read that some people are having problems with Python packages with the M1. Let me see if I can get a M1 instance on one of the cloud providers to test. I'll check this a bit later today

**Reply by haltakov** [reply]
I'm struggling getting an M1 instance online... Can you try to first run `pip install Jinja2` and then try to install simple-photo-gallery?

I'll now build a test version without specific version pinning, since the exact versions are not so important.

**Reply by haltakov** [reply]
OK, I created a test version that has no requirements on the versions at all. You can try to install it in a fresh environment.
```
pip install simple-photo-gallery==1.5.1b1
```

**Reply by PseudoSurfer** [reply]
This is what I get:

Collecting simple-photo-gallery==1.5.1b1
  Downloading simple_photo_gallery-1.5.1b1-py3-none-any.whl (2.5 MB)
     |████████████████████████████████| 2.5 MB 3.9 MB/s 
Requirement already satisfied: pillow in ./.local/lib/python3.7/site-packages (from simple-photo-gallery==1.5.1b1) (8.0.1)
Requirement already satisfied: requests in ./.local/lib/python3.7/site-packages (from simple-photo-gallery==1.5.1b1) (2.25.1)
Collecting jinja2
  Using cached Jinja2-2.11.3-py2.py3-none-any.whl (125 kB)
Collecting opencv-python
  Using cached opencv-python-4.5.1.48.tar.gz (88.3 MB)
  Installing build dependencies ... error

then it cycles through each version until

Using cached opencv-python-3.4.10.37.tar.gz (87.5 MB)
  Installing build dependencies ... error

But I have this installed through homebrew:

miniforge was successfully installed!
luv@Justins-Mini ~ % brew install numpy
Warning: numpy 1.20.2 is already installed and up-to-date.
To reinstall 1.20.2, run:
  brew reinstall numpy
luv@Justins-Mini ~ % brew install opencv
Warning: opencv 4.5.2 is already installed and up-to-date.
To reinstall 4.5.2, run:
  brew reinstall opencv
luv@Justins-Mini ~ % 

So my question is why doesn't your script find the version I have installed? 

Trying again I used conda and installed opencv again (but to be honest I've no idea what I'm doing. I just search, copy, paste, and try again.) 

luv@Justins-Mini ~ % python -c 'import cv2'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'cv2'
luv@Justins-Mini ~ % python -c 'import cv4'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'cv4'

So now I'm not sure but I'll keep trying

**Reply by PseudoSurfer** [reply]
I notice that it starts from 4.5.1 which is the latest version but my system says it has 4.5.2 installed. I'm not sure how I got 4.5.2 but it says it's there.

**Reply by PseudoSurfer** [reply]
It's from homebrew https://formulae.brew.sh/formula/opencv#default

**Comment by haltakov** [comment]
Sorry I was trying to get access to an M1, but this Is not easy… Can you maybe try to uninstall all dependencies with pip uninstall (opencv-python and pillow) and simple-photo-gallery and then run int again.

Another thing to try would be to remove opencv that you installed with home-brew and let simple-photo-gallery install opencv-python by itself.

**Reply by Bennewith** [reply]
Thanks for getting back to me. After loads of trial and error (specifically trying to build opencv using these directions: 
https://sayak.dev/install-opencv-m1/ which still failed. Uninstalled and reinstalled X-Code too)
I finally got opencv installed using this command: 
conda install -c conda-forge opencv   from here:  https://anaconda.org/conda-forge/opencv
But when I then try: pip install simple-photo-gallery==1.5.1b1 It still tries to find and install opencv 4.5.1 when 4.5.2 got installed for a second time in my conda env. 
Any ideas?

**Reply by Bennewith** [reply]
In reference to your suggestion above, I've created and removed my miniforge conda env 4 or 5 times and reinstall python and numpy each time before trying to get opencv installed.

**Reply by Bennewith** [reply]
> conda list
# packages in environment at /opt/homebrew/Caskroom/miniforge/base:
#
# Name                    Version                   Build  Channel
brotlipy                  0.7.0           py39h46acfd9_1001    conda-forge
bzip2                     1.0.8                h3422bc3_4    conda-forge
c-ares                    1.17.1               h27ca646_1    conda-forge
ca-certificates           2020.12.5            h4653dfc_0    conda-forge
cairo                     1.16.0            he69dfd1_1008    conda-forge
certifi                   2020.12.5        py39h2804cbe_1    conda-forge
cffi                      1.14.5           py39h702c04f_0    conda-forge
chardet                   4.0.0            py39h2804cbe_1    conda-forge
conda                     4.10.1           py39h2804cbe_0    conda-forge
conda-package-handling    1.7.3            py39h5161555_0    conda-forge
cryptography              3.4.7            py39h73257c9_0    conda-forge
ffmpeg                    4.3.1                h22aafd7_2    conda-forge
fontconfig                2.13.1            heb65262_1005    conda-forge
freetype                  2.10.4               h17b34a0_1    conda-forge
gettext                   0.19.8.1          hea66d9f_1005    conda-forge
gmp                       6.2.1                h9f76cd9_0    conda-forge
gnutls                    3.6.13               h706517b_1    conda-forge
graphite2                 1.3.13            h9f76cd9_1001    conda-forge
harfbuzz                  2.8.0                h2a75196_1    conda-forge
hdf5                      1.10.6          nompi_h0fc092c_1114    conda-forge
icu                       68.1                 h17758a7_0    conda-forge
idna                      2.10               pyh9f0ad1d_0    conda-forge
jasper                    2.0.14               h9af5a26_2    conda-forge
jpeg                      9d                   h27ca646_0    conda-forge
krb5                      1.17.2               h17618d6_0    conda-forge
lame                      3.100             h27ca646_1001    conda-forge
libblas                   3.9.0                8_openblas    conda-forge
libcblas                  3.9.0                8_openblas    conda-forge
libcurl                   7.76.1               h0ebd9f5_1    conda-forge
libcxx                    11.1.0               h168391b_0    conda-forge
libedit                   3.1.20191231         hc8eb9b7_2    conda-forge
libev                     4.33                 h642e427_1    conda-forge
libffi                    3.3                  h9f76cd9_2    conda-forge
libgfortran               5.0.0.dev0      11_0_1_hf114ba7_22    conda-forge
libgfortran5              11.0.1.dev0         hf114ba7_22    conda-forge
libglib                   2.68.1               h6c200e4_0    conda-forge
libiconv                  1.16                 h642e427_0    conda-forge
liblapack                 3.9.0                8_openblas    conda-forge
liblapacke                3.9.0                8_openblas    conda-forge
libnghttp2                1.43.0               hf3018f0_0    conda-forge
libopenblas               0.3.12          openmp_h2ecc587_1    conda-forge
libopencv                 4.5.1            py39h9e916e0_1    conda-forge
libpng                    1.6.37               hf7e6567_2    conda-forge
libprotobuf               3.15.8               hccf11d3_0    conda-forge
libssh2                   1.9.0                hb80f160_6    conda-forge
libtiff                   4.2.0                hcbb2f27_1    conda-forge
libwebp-base              1.2.0                h27ca646_2    conda-forge
libxml2                   2.9.10               h538f51a_4    conda-forge
llvm-openmp               11.1.0               hf3c4609_1    conda-forge
lz4-c                     1.9.3                h9f76cd9_0    conda-forge
ncurses                   6.2                  h9aa5885_4    conda-forge
nettle                    3.6                  hc6a1b29_0    conda-forge
numpy                     1.20.2           py39h1f3b974_0    conda-forge
openh264                  2.1.1                habe5f53_0    conda-forge
openssl                   1.1.1k               h27ca646_0    conda-forge
pcre                      8.44                 hb904e53_0    conda-forge
pip                       21.1               pyhd8ed1ab_0    conda-forge
pixman                    0.40.0               h27ca646_0    conda-forge
py-opencv                 4.5.1            py39hfa6204d_1    conda-forge
pycosat                   0.6.3           py39h46acfd9_1006    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py39h2804cbe_3    conda-forge
python                    3.9.2           hcbd9b3a_0_cpython    conda-forge
python_abi                3.9                      1_cp39    conda-forge
readline                  8.1                  hedafd6a_0    conda-forge
requests                  2.25.1             pyhd3deb0d_0    conda-forge
ruamel_yaml               0.15.80         py39h46acfd9_1004    conda-forge
setuptools                49.6.0           py39h2804cbe_3    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sqlite                    3.35.5               hc49ca36_0    conda-forge
tk                        8.6.10               hf7e6567_1    conda-forge
tqdm                      4.60.0             pyhd8ed1ab_0    conda-forge
tzdata                    2021a                he74cb21_0    conda-forge
urllib3                   1.26.4             pyhd8ed1ab_0    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
x264                      1!161.3030           h27ca646_0    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
yaml                      0.2.5                h642e427_0    conda-forge
zlib                      1.2.11            h31e879b_1009    conda-forge
zstd                      1.4.9                h5b28eab_0    conda-forge

**Comment by haltakov** [comment]
I'm now trying it out as well... I did a fresh install of 1.5.1b

```
(.venv) (base)  ~/Projects/test_spg_m1  pip install simple-photo-gallery==1.5.1b1
Collecting simple-photo-gallery==1.5.1b1
  Using cached simple_photo_gallery-1.5.1b1-py3-none-any.whl (2.5 MB)
Collecting opencv-python
  Downloading opencv_python-4.5.2.52-cp36-cp36m-macosx_10_15_x86_64.whl (43.6 MB)
     |████████████████████████████████| 43.6 MB 10.8 MB/s 

```

**Comment by haltakov** [comment]
It installs opencv 4.5.2, but from the opencv-python package...

**Comment by haltakov** [comment]
OK, I built a new version for you removing the opencv requirement. Hopefully it will work when you install it yourself.
```
pip install simple-photo-gallery==1.5.1b2
```

**Reply by Bennewith** [reply]
> cd Photo_Galleries
> gallery init -p https://photos.app.goo.gl/CZYFLYeuNNpi1QPA9
zsh: command not found: gallery
> conda activate dev
> gallery init -p https://photos.app.goo.gl/CZYFLYeuNNpi1QPA9
zsh: command not found: gallery
> cd ~
> cd /opt/homebrew/Caskroom/miniforge/base/envs/dev
> conda activate dev
> gallery init -p https://photos.app.goo.gl/CZYFLYeuNNpi1QPA9
zsh: command not found: gallery

Admittedly I don't know what I'm doing with these python enviroments (I'm not sure if I need to create my galleries within the enviroment or not) So I tried the way that it worked before and also within the enviroment folder but niether of them worked.

**Reply by Bennewith** [reply]
The install seemed to go fine when I used the new version.

**Reply by Bennewith** [reply]
I read somewhere that packages installed with pip don't play nice with packages installed with Conda.

> pip freeze
certifi==2020.12.5
chardet==4.0.0
idna==2.10
Jinja2==3.0.0
MarkupSafe==2.0.0
numpy @ file:///Users/runner/miniforge3/conda-bld/numpy_1616933194735/work
Pillow==8.2.0
requests==2.25.1
selenium==3.141.0
simple-photo-gallery==1.5.1b2
urllib3==1.26.4

**Reply by Bennewith** [reply]
I installed simple photo gallery directly into the Miniforge Conda enviroment too

/opt/homebrew/Caskroom/miniforge/base/envs/dev/bin/pip install simple-photo-gallery==1.5.1b2
Requirement already satisfied: simple-photo-gallery==1.5.1b2 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (1.5.1b2)
Requirement already satisfied: selenium in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from simple-photo-gallery==1.5.1b2) (3.141.0)
Requirement already satisfied: requests in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from simple-photo-gallery==1.5.1b2) (2.25.1)
Requirement already satisfied: jinja2 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from simple-photo-gallery==1.5.1b2) (3.0.0)
Requirement already satisfied: pillow in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from simple-photo-gallery==1.5.1b2) (8.2.0)
Requirement already satisfied: MarkupSafe>=2.0.0rc2 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from jinja2->simple-photo-gallery==1.5.1b2) (2.0.0)
Requirement already satisfied: chardet<5,>=3.0.2 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from requests->simple-photo-gallery==1.5.1b2) (4.0.0)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from requests->simple-photo-gallery==1.5.1b2) (1.26.4)
Requirement already satisfied: idna<3,>=2.5 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from requests->simple-photo-gallery==1.5.1b2) (2.10)
Requirement already satisfied: certifi>=2017.4.17 in /opt/homebrew/Caskroom/miniforge/base/envs/dev/lib/python3.8/site-packages (from requests->simple-photo-gallery==1.5.1b2) (2020.12.5)

**Reply by haltakov** [reply]
Ok, so now the installation finishes successfully right?

You should call the gallery using `gallery-init` not `gallery init` ;)

**Reply by Bennewith** [reply]
Success!!! Thanks again for everything.

**Reply by Bennewith** [reply]
To be clear in case you want to post instructions or other info. I'm using Miniforge Conda and used pip inside the env to install Simple Photo Gallery in the env. It doesnt appear as though the homebrew installed packages are interfering or co-operating with the Conda env at all.

**Reply by haltakov** [reply]
OK, thanks for the information! I'm happy we found a solution that works for you, but I need to find a Mac with an M1 to find a proper solution...

**Reply by haltakov** [reply]
Thanks for testing this and for your patience :)

**Reply by PseudoSurfer** [reply]
Hello
Is there any reason that it limits a gallery to 30 images? I have a few that are more than 30 including one that is 127 images.

**Reply by haltakov** [reply]
Hmm, there is no limit to 30. I have galleries with hundreds of photos. What error do you get?

**Reply by PseudoSurfer** [reply]
So I broke it into sections as described 



  {{ gallery_macros.section(0, 31,
                              'Arriving in Honduras',
                              'Departing San Pedro Sula and arriving on the island of Roatan',
                              images)}}

  {{ gallery_macros.section(31, 70,
                              'Island Hopping',
                              'Departing the island of Roatan and arriving on the island of Utila',
                              images)}}

  {{ gallery_macros.section(70, 97,
                              'Exploring Utila',
                              'The reason for the trip was to put my dads ashes on Utila',
                              images)}}

  {{ gallery_macros.section(97, 126,
                              'Leaving',
                              'Leaving Roatan and heading back to San Pedro Sula',
                              images)}}

  

All photos processed!
The image descriptions are stored in images_data.json. You can edit the file to add more descriptions and build the gallery again.
Creating the index.html...
Something went wrong while generating the gallery HTML: list object has no element 30

**Reply by PseudoSurfer** [reply]
As far as an error goes when trying galleries with more than 30 images, it doesn't give an error, it just stops at 30.

**Reply by PseudoSurfer** [reply]
this was the info at the beginning
gallery-build
Building the Simple Photo Gallery...
Generating thumbnails...
Generating the images_data.json file...
Starting Firefox webdriver...
Loading album from https://photos.app.goo.gl/d99MskBz2AXrRFfJ8...
Finding photos...
Photos found: 30

**Reply by haltakov** [reply]
OK, I see what's wrong... The problem is that the script that downloads photos from Google only finds the first 30 photos. This is because Google only initializes 30 photos and loads the rest dynamically while scrolling the album. This seems to be a new behavior in Google Photos...

I can try to work around this, but it may not be very easy. Alternatively, you could try downloading the photos locally and creating a gallery in this way (only `gallery-init` from the folders where the photos are). Will this work for you?

**Reply by PseudoSurfer** [reply]
No not really. I have already broken up into albums of no more than 30 images each. I'll just copy and paste them together in the HTML file. Thanks again for everything
